modules/ModuleNewsletterUnsubscribeNotificationCenter.php

```diff
@@ -0,0 +1,246 @@
+        }
+
+        $this->Template->selectedChannels = $arrChannels;
+
+        // Check if there are any new subscriptions
+        $arrSubscriptions = array();
+
+        if (($objSubscription = \NewsletterRecipientsModel::findBy(array("email=? AND active=1"), $varInput)) !== null)
+        {
+            $arrSubscriptions = $objSubscription->fetchEach('pid');
+        }
+
+        $arrRemove = array_intersect($arrChannels, $arrSubscriptions);
+
+        if (empty($arrRemove) || !\is_array($arrRemove))
+        {
+            $this->Template->mclass = 'error';
+            $this->Template->message = $GLOBALS['TL_LANG']['ERR']['unsubscribed'];
+
+            return false;
+        }
+
+        // Validate the captcha
+        if ($objWidget !== null)
+        {
+            $objWidget->validate();
+
+            if ($objWidget->hasErrors())
+            {
+                return false;
+            }
+        }
+
+        return array($varInput, $arrRemove);
+    }
+
+    /**
+     * Remove the recipient
+     *
+     * @param string $strEmail
+     * @param array  $arrRemove
+     */
+    protected function removeRecipient($strEmail, $arrRemove)
+    {
+        // Remove the subscriptions
+        if (($objRemove = \NewsletterRecipientsModel::findByEmailAndPids($strEmail, $arrRemove)) !== null)
+        {
+            while ($objRemove->next())
+            {
+                $strHash = md5($objRemove->email);
+
+                // Add a blacklist entry (see #4999)
+                if (version_compare(VERSION, '4.1', '>=') && $objBlacklist = (\NewsletterBlacklistModel::findByHashAndPid($strHash, $objRemove->pid)) === null)
+                {
+                    $objBlacklist = new \NewsletterBlacklistModel();
+                    $objBlacklist->pid = $objRemove->pid;
+                    $objBlacklist->hash = $strHash;
+                    $objBlacklist->save();
+                }
+
+                $objRemove->delete();
+            }
+        }
+
+        // HOOK: post unsubscribe callback
+        if (isset($GLOBALS['TL_HOOKS']['removeRecipient']) && \is_array($GLOBALS['TL_HOOKS']['removeRecipient']))
+        {
+            foreach ($GLOBALS['TL_HOOKS']['removeRecipient'] as $callback)
+            {
+                $this->import($callback[0]);
+                $this->{$callback[0]}->{$callback[1]}($strEmail, $arrRemove);
+            }
+        }
+
+        $this->sendNotification($strEmail, $arrRemove);
+        $this->redirectToJumpToPage();
+
+        if (version_compare(VERSION, '4.1', '>='))
+        {
+            \System::getContainer()->get('session')->getFlashBag()->set('nl_removed', $GLOBALS['TL_LANG']['MSC']['nl_removed']);
+        }
+        else
+        {
+            $_SESSION['UNSUBSCRIBE_CONFIRM'] = $GLOBALS['TL_LANG']['MSC']['nl_removed'];
+        }
+
+        $this->reload();
+    }
+
+    protected function compileConfirmationMessage()
+    {
+        if (version_compare(VERSION, '4.1', '>='))
```

Comment:
I think we could raise the minimum dependencies to Contao 3.5 or 4.4, all the others are unmaintained anyway, wdyt @aschempp ?

Reply:
we could require 3.5, but what advantage would that be? I'd rather keep it as is and then rewrite at some point for Contao 4...